base/asyncmap.jl

```diff
@@ -19,44 +19,97 @@ type AsyncCollector
     f
     results
     enumerator::Enumerate
-    ntasks::Int
+    max_tasks::Function
+    c::Channel{Tuple{Int, Any}}     # to communicate with the tasks
+
+    AsyncCollector(f, r, en::Enumerate, mt::Function, c::Channel) = new(f, r, en, mt, c)
 end
 
 function AsyncCollector(f, results, c...; ntasks=0)
     if ntasks == 0
         ntasks = max(nworkers(), 100)
+        max_tasks = ()->ntasks
+    elseif isa(ntasks, Integer)
+        max_tasks = ()->ntasks
+    elseif isa(ntasks, Function)
+        max_tasks = ntasks
+    else
+        throw(ArgumentError("ntasks must be an Integer or a zero-arg function returning the maximum number of tasks allowed."))
```

Comment:
Allows for the number of concurrent tasks in `AsyncCollector` to be dynamic. One of the requested features was that pmap must be able to leverage workers as and when they come online. This allows for a user of `AsyncCollector `, like `pmap` to increase count of parallel tasks depending on current number of workers. Note that it is not a major issue if the number of tasks is more than the number of workers, but it will be quite inefficient the other way around.